### PR TITLE
[KYUUBI #1351] Implement api: /${version}/sessions/${identifier}/operations

### PR DESCRIPTION
**Pull request by simon824**
### _Why are the changes needed?_
This is a subtask of umbrella issue #KPIP-1

- /${version}/sessions/${identifier}/operations
  - mapping:
    - ICLIService#executeStatement
    - ICLIService#executeStatementAsync
    - ICLIService#getTypeInfo
    - ICLIService#getCatelogs
    - ICLIService#getSchemas
    - ICLIService#getTables
    - ICLIService#getTableTypes
    - ICLIService#getColumns
    - ICLIService#getFunctions
  - desc: create a new operation under a session
  - method: POST
  - params:
  - operation: e.g. GetTableTypes
  - addition: required params to create an Operation
  - returns: an instance of OperationHandle

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by simon824** [comment]
cc @yanghua @pan3793 @yaooqinn

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala [hunk 1]
```diff
@@ -207,4 +208,40 @@ class SessionsResourceSuite extends KyuubiFunSuite {
         assert(404 == response.getStatus)
     }
   }
+
+
+  test("test get operations") {
+    val requestObj = SessionOpenRequest(
+      1, "admin", "123456", "localhost", Map("testConfig" -> "testValue"))
+
+    RestFrontendServiceSuite.withKyuubiRestServer {
```
Comment (pan3793) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/SessionsResourceSuite.scala:217]: #1331 refactor the test suite, please rebase master

**Comment by yaooqinn** [comment]
/${version}/sessions/${identifier}/operations
desc: get all the operation list hosted a specific session binding via an identifier
method: GET
param:identifier instance of SessionHandle
returns: a list of Operation

**Comment by yaooqinn** [comment]
> /${version}/sessions/${identifier}/operations desc: get all the operation list hosted a specific session binding via an identifier method: GET param:identifier instance of SessionHandle returns: a list of Operation

This PR seems not correct, according to the KPIP-1. 

And this intention of this API seems to expose `opHandleSet` in a session, which is actually a private member? What's the motivation of this?

**Comment by simon824** [comment]
Not this one. 

> /${version}/sessions/${identifier}/operations desc: get all the operation list hosted a specific session binding via an identifier method: GET param:identifier instance of SessionHandle returns: a list of Operation


- /${version}/sessions/${identifier}/operations
  - mapping:
    - ICLIService#executeStatement
    - ICLIService#executeStatementAsync
    - ICLIService#getTypeInfo
    - ICLIService#getCatelogs
    - ICLIService#getSchemas
    - ICLIService#getTables
    - ICLIService#getTableTypes
    - ICLIService#getColumns
    - ICLIService#getFunctions
  - desc: create a new operation under a session
  - method: POST
  - params:
  - operation: e.g. GetTableTypes
  - addition: required params to create an Operation
  - returns: an instance of OperationHandle

**Comment by yaooqinn** [comment]
backendService

> Not this one.
> 
> > /${version}/sessions/${identifier}/operations desc: get all the operation list hosted a specific session binding via an identifier method: GET param:identifier instance of SessionHandle returns: a list of Operation
> 
> * /${version}/sessions/${identifier}/operations
>   
>   * mapping:
>     
>     * ICLIService#executeStatement
>     * ICLIService#executeStatementAsync
>     * ICLIService#getTypeInfo
>     * ICLIService#getCatelogs
>     * ICLIService#getSchemas
>     * ICLIService#getTables
>     * ICLIService#getTableTypes
>     * ICLIService#getColumns
>     * ICLIService#getFunctions
>   * desc: create a new operation under a session
>   * method: POST
>   * params:
>   * operation: e.g. GetTableTypes
>   * addition: required params to create an Operation
>   * returns: an instance of OperationHandle

we shall implement separately, like the `def executeStatement` where we call `backendService.executeStatement`

**Comment by yaooqinn** [comment]
for `executeStatement`, do we need to get the actual final result？with FetchResult API?

**Comment by simon824** [comment]
> executeStatement

Do you mean one API corresponds to one method?
like   /${version}/sessions/${identifier}/executeStatement -> ICLIService#executeStatement ?

**Comment by simon824** [comment]
> for `executeStatement`, do we need to get the actual final result？with FetchResult API?

Yes, we need `FetchResults` API to get the actual final result.

**Comment by yaooqinn** [comment]
> > executeStatement
> 
> Do you mean one API corresponds to one method? like /${version}/sessions/${identifier}/executeStatement -> ICLIService#executeStatement ?

yes

**Comment by simon824** [comment]
cc @yaooqinn @yanghua

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 2]
```diff
@@ -109,6 +110,125 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+      backendService.executeStatement(sessionHandle, statement, runAsync.toBoolean,
+        queryTimeout.toLong)
+    } catch {
+      case NonFatal(_) =>
+        throw new NotFoundException(s"Error executing statement")
+    }
+  }
+
+  @GET
+  @Path("{sessionHandle}/getTypeInfo")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:131]: This naming is not suited for the rule of the RESTful API.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 3]
```diff
@@ -109,6 +110,125 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+    try {
+      backendService.getTypeInfo(sessionHandle)
+    } catch {
+      case NonFatal(_) =>
+        throw new NotFoundException(s"Error getting typeInfo")
+    }
+  }
+
+  @GET
+  @Path("{sessionHandle}/getCatalogs")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:143]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 4]
```diff
@@ -109,6 +110,125 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+    try {
+      backendService.getCatalogs(sessionHandle)
+    } catch {
+      case NonFatal(_) =>
+        throw new NotFoundException(s"Error getting catalogs")
+    }
+  }
+
+  @GET
+  @Path("{sessionHandle}/getSchemas")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:155]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 5]
```diff
@@ -109,6 +110,125 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+        throw new NotFoundException(s"Error executing statement")
+    }
+  }
+
+  @GET
+  @Path("{sessionHandle}/getTypeInfo")
+  def getTypeInfo(@PathParam("sessionHandle") sessionHandleStr: String): OperationHandle = {
+    val sessionHandle = getSessionHandle(sessionHandleStr)
+    try {
+      backendService.getTypeInfo(sessionHandle)
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:135]: Actually, this behavior is to create or submit a `getTypeInfo` operation, right?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 6]
```diff
@@ -109,6 +110,125 @@ private[v1] class SessionsResource extends ApiRequestContext {
     Response.ok().build()
   }
 
+  @GET
+  @Path("{sessionHandle}/executeStatement")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:114]: ditto

**Comment by yanghua** [comment]
> > for `executeStatement`, do we need to get the actual final result？with FetchResult API?
> 
> Yes, we need `FetchResults` API to get the actual final result.

IMO, fetch API should be `GET`, while `Session#getCatalogs` should be a post request to create an operation and return an `operation handle`.

WDYT? @simon824 @yaooqinn

**Comment by simon824** [comment]
> > > for `executeStatement`, do we need to get the actual final result？with FetchResult API?
> > 
> > 
> > Yes, we need `FetchResults` API to get the actual final result.
> 
> IMO, fetch API should be `GET`, while `Session#getCatalogs` should be a post request to create an operation and return an `operation handle`.
> 
> WDYT? @simon824 @yaooqinn

Make sense.

**Comment by simon824** [comment]
This API is actually to create different type of OperationHandle. From this perspective, I think it’s better to implement all the methods in the same API like we designed before. WDYT @yaooqinn

**Comment by yaooqinn** [comment]
> This API is actually to create different type of OperationHandle?

which API?

**Comment by simon824** [comment]
> > This API is actually to create different type of OperationHandle?
> 
> which API?

- /${version}/sessions/${identifier}/operations
  - mapping:
    - ICLIService#executeStatement
    - ICLIService#executeStatementAsync
    - ICLIService#getTypeInfo
    - ICLIService#getCatelogs
    - ICLIService#getSchemas
    - ICLIService#getTables
    - ICLIService#getTableTypes
    - ICLIService#getColumns
    - ICLIService#getFunctions
  - desc: create a new operation under a session
  - method: POST

**Comment by yaooqinn** [comment]
Take GetCatalogs as an example.

`REST Client` _POST_ `GetCatalogsReq(sessionHandle)` -> `REST FE` -> call -> Kyuubi BE `getCatalogs` -> `GetCatalogsResp(operationHandle, OK/ERROR)` -> `REST Client`

`REST Client`  _GET_`FetchResultsReq(operationHandle, orientation, maxRows, fetchType(log/result))` -> `REST FE` -> call -> Kyuubi BE `fetchResults` -> `FetchResultsResp(OK/ERROR, hasMoreRows, resultset)` -> `REST Client` -> if `hasMoreRows` -> repeat


FYI, https://prestodb.io/docs/current/develop/client-protocol.html

**Comment by yanghua** [comment]
> Take GetCatalogs as an example.
> 
> `REST Client` _POST_ `GetCatalogsReq(sessionHandle)` -> `REST FE` -> call -> Kyuubi BE `getCatalogs` -> `GetCatalogsResp(operationHandle, OK/ERROR)` -> `REST Client`
> 
> `REST Client` _GET_`FetchResultsReq(operationHandle, orientation, maxRows, fetchType(log/result))` -> `REST FE` -> call -> Kyuubi BE `fetchResults` -> `FetchResultsResp(OK/ERROR, hasMoreRows, resultset)` -> `REST Client` -> if `hasMoreRows` -> repeat
> 
> FYI, https://prestodb.io/docs/current/develop/client-protocol.html

+1

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 7]
```diff
@@ -109,6 +109,111 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+    try {
+      backendService.getTypeInfo(sessionHandle)
+    } catch {
+      case NonFatal(_) =>
+        throw new NotFoundException(s"Error getting typeInfo")
+    }
+  }
+
+  @POST
+  @Path("{sessionHandle}/getcatalogs")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:139]: I mean we should not contain verb e.g. `get` in the RESTful URI. The "R" in rest means resource, it is recommended to be a noun.

Comment (simon824) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:139]: What about `catalogs` or `catelogshandle`?

Comment (yanghua) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:139]: Actually, this API creats a `getCatalogs` operation under a specific session. So, what about renaming to be `/sessions/{sessionHandle}/operations/catalogs`?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 8]
```diff
@@ -109,6 +109,111 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+
+  @POST
+  @Path("{sessionHandle}/operations/tabletypes")
+  def getTableTypes(@PathParam("sessionHandle") sessionHandleStr: String): OperationHandle = {
+    val sessionHandle = getSessionHandle(sessionHandleStr)
+    try {
+      backendService.getTableTypes(sessionHandle)
+    } catch {
+      case NonFatal(_) =>
+        throw new NotFoundException(s"Error getting tableTypes")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:185]: ditto

**Review by yanghua — COMMENTED** [review]
@simon824 Thanks, left some comments you can consider.

**Review by yanghua — APPROVED** [review]
LGTM

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thank you @simon824 @yanghua for the work !

merging to master. FYI @yaooqinn

**Pull request closed without merging**